Deviseの課題を提出させていただきます。

kotokotodonabe opened the pull request.

Commit:
deviseのセットアップとログイン画面をrootに変更

Commit:
プロフィール編集機能の実装

Commit:
ログインユーザーのプロフィールとその他のユーザーのコントローラーを分けた

【理由】
1.Railsのアプリケーションでは、ユーザーの詳細ページを表示させる際はshowアクションの方が自然
2.分けた方が管理しやすい

Commit:
ページング処理の追加

Commit:
letter_opener_webを有効化し、パスワード再設定機能の実装完了

【letter_opener_webとは】
メールを送信した際に、ブラウザ上でそのメールの内容を確認できるgem。

【なぜ採用したのか】
パスワード再設定の機能を実装しようとした際に、gmailでパスワード再設定のメールを受け取るには
色々設定する必要があったのと、パスワード再設定機能を実現するにあたり一々メールを開いて作業を
進めるのは効率が悪いと考えたから。

Commit:
プロフィールページからパスワード変更する機能の実装

update_without_current_passwordというメソッドをmodels>user.rbに定義して、パスワードの
値が送られてきた場合更新、送られてこなかった場合更新しない処理を行なった。

Commit:
I18nで国際化とroutingの修正

Commit:
update時のルートを変更

Commit:
新規登録時のエラーメッセージを日本語化

Commit:
rubocopとlintを通した

maedana commented on Gemfile:
```diff
@@ -61,4 +63,7 @@ gem 'net-pop'
 gem 'carrierwave'
+gem 'devise'
 gem 'kaminari'
+gem 'pry'
```
デバッグ用ですよね。productionには不要だと思いますし、development groupのみにするなどインストールされるタイミングは限定しましょう。

maedana commented on app/assets/stylesheets/application.css:
```diff
@@ -13,3 +13,6 @@
  */
+body {
+  height: 100vh;
+}
```
なんのために入れてます? 全体の高さ固定したいのでしょうか?

maedana commented on app/assets/stylesheets/my_profile.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the my_profile controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
自動生成の無意味なコードはコミットしないようにしましょうー。こういう話ですね。
https://bootcamp.fjord.jp/pages/info-for-github#meaningless-file

maedana replied:
他にもhelperなど色々自動生成しただけのものがありそうなので、それぞれ確認して消しましょう。提出条件のその他欄にも`プルリクエストには課題の要件満たすために必要なdiffだけ発生している状態がベスト`とありますしね。

maedana commented on config/routes.rb:
```diff
@@ -1,4 +1,27 @@
+    post "users/login", to: "users/sessions#create"
+    delete "users/logout", to: "users/sessions#destroy"
+    get "users/my_profile/edit", to: "users/registrations#edit"
+    patch "users/my_profile/edit", to: "users/registrations#update"
```
devise標準のroutingが使いたくないのでしょうか? 拘りがないなら標準のままでいいんじゃないかと思いますがいかがでしょうか?

maedana commented on config/routes.rb:
```diff
@@ -1,4 +1,27 @@
+  get  'users/show'
+  get  'users', to: 'users#index'
+  # ユーザー詳細ページへのルーティング
+  match 'users/:id' => 'users#show', :via => :get
```
```ruby
  get 'users/my_profile', to: 'my_profile#show'
  get  'users/show'
  get  'users', to: 'users#index'
  # ユーザー詳細ページへのルーティング
  match 'users/:id' => 'users#show', :via => :get
```

のように独自routingは理由がなければ使わずに、なるべく`resource/resources`を使いましょう。少し上の`resources :books`のようなことですね。これではアプリが成長して数が増えていくメンテナンスが辛くなっていくかなーと思います。

maedana commented on app/controllers/my_profile_controller.rb:
```diff
@@ -0,0 +1,7 @@
+
+class MyProfileController < ApplicationController
+  def show
+    @user = current_user
```
インスタンス変数にせずとも`current_user`のままviewで使えばいいんじゃないですかね? その方がリーダブルですし。

maedana commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  # def after_confirmation_path_for(resource_name, resource)
+  #   super(resource_name, resource)
+  # end
+end
```
カスタマイズしてないですし、不要かなーと思います。routingも合わせて見直しましょう。devise標準のcontrollerをカスタマイズするときのみ必要なものですね。

参考
https://github.com/heartcombo/devise#configuring-controllers

maedana commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  # def after_omniauth_failure_path_for(scope)
+  #   super(scope)
+  # end
+end
```
同上です。そもそも`OmniauthCallbacksController`が何のためのcontrollerなのか調べてみるといいかなーと思います。たぶんどういうものかわからず書いているのかなって思いました。

maedana commented on app/controllers/users/passwords_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  # def after_sending_reset_password_instructions_path_for(resource_name)
+  #   super(resource_name)
+  # end
+end
```
同上です。

maedana replied:
何もカスタマイズしてないのでcontroller自体が不要ですねー。`routes.rb`も合わせて不要なところ削りましょう

maedana commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,78 @@
+  # def after_inactive_sign_up_path_for(resource)
+  #   super(resource)
+  # end
+end
```
これは独自カスタマイズしてますね。こういうのがdeviseのcontrollerのカスタマイズのときの使い方ですねー。

maedana commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  # def configure_sign_in_params
+  #   devise_parameter_sanitizer.permit(:sign_in, keys: [:attribute])
+  # end
+end
```
同上です。

maedana commented on app/controllers/users/unlocks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  # def after_unlock_path_for(resource)
+  #   super(resource)
+  # end
+end
```
同上です。

maedana commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,11 @@
+
+class UsersController < ApplicationController
+  def index
+    @user = User.where.not(id: current_user.id).order(:remember_created_at, :id).page(params[:page]).per(3)
```
`remember_created_at`がorderに入ってるのが気になります。どういう意図なんでしょう?

kotokotodonabe replied:
本来であればcreated_atを記述するつもりだったのですが、誤ってしました🙇

maedana commented on app/models/user.rb:
```diff
@@ -0,0 +1,18 @@
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
controllerの所でいくつか指摘しましたけども、ここでそもそも`omniauthable`とか`lockable`のような機能が有効になってないですし、この時点で一部のviewやcontrollerは不使用ということがわかりますねー。

maedana commented on app/views/books/index.html.erb:
```diff
@@ -1,4 +1,4 @@
-<p id="notice"><%= notice %></p>
+<%# <p id="notice"><%= notice %></p> %>
```
不要ならコメントアウトではなく消しましょうー

maedana commented on app/views/users/confirmations/new.html.erb:
```diff
@@ -0,0 +1,16 @@
+  </div>
+<% end %>
+
+<%= render "users/shared/links" %>
```
deviseの未使用viewですし消しましょうー。他にも複数あります。それぞれ確認して消しましょう。

maedana commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,16 @@
+
+<% @user.each do |user| %>
+  <ul class="user-list">
+    <a href="/users/<%= user.id %>" class="user-link">
```
`link_to`ヘルパー使って書きましょう。またURLヘルパーも使いましょう。少し上の`users_path`みたいなものですね。`routes.rb`で`resource/resources`使うようにすることで詳細ページでもヘルパーが生成されるかと。

maedana commented on app/views/users/mailer/password_change.html.erb:
```diff
@@ -0,0 +1,3 @@
+<p>Hello <%= @resource.email %>!</p>
+
+<p>We're contacting you to notify you that your password has been changed.</p>
```
パスワード忘れ関連ですがi18nでの日本語化がされてないですねー。他にも対応できてないviewが複数ありそうです。繰り返しですが不要なviewは消す。使っているviewはi18nでの日本語化をちゃんとやりましょう。

maedana commented on config/application.rb:
```diff
@@ -10,6 +10,10 @@ module BooksApp
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 6.1
+    config.time_zone = 'Tokyo'
+    config.active_record.default_timezone = :local
```
これは何ために書いているか理解できてますか? また必要性を説明できますか?

kotokotodonabe replied:
@maedana 
すいません、なんとなく入れてしまっていただけでした。

maedana replied:
> すいません、なんとなく入れてしまっていただけでした。

書いたコードは一字一句説明出来るのが理想です。理解せずに書くことはなるべく避けるのが望ましいですね

maedana commented on config/application.rb:
```diff
@@ -10,6 +10,10 @@ module BooksApp
+    config.time_zone = 'Tokyo'
+    config.active_record.default_timezone = :local
+    config.i18n.default_locale = :ja
+    config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}').to_s]
```
同上です。これは何ために書いているか理解できてますか? また必要性を説明できますか?

maedana commented on config/locales/devise.en.yml:
```diff
@@ -0,0 +1,65 @@
+      not_locked: "was not locked"
+      not_saved:
+        one: "1 error prohibited this %{resource} from being saved:"
+        other: "%{count} errors prohibited this %{resource} from being saved:"
```
これはdeviseのi18nの英語ファイルですが、日本語版にあたるものが見当たらないのでdeviseの各メッセージが国際化されてないのではないかと思うのですがいかがでしょうか?

と思ったら
config/locales/model/ja.yml
で一部だけ入れてあるんですね。一部だけでなく、
config/locales/devise.en.yml に対応した config/locales/devise.ja.yml があって全てのメッセージの国際化がいるかなーと思います。で、それを自力でやるのは大変なので、ヒントに
> Devise画面の日本語訳はgemを利用すると便利

とあったりします。どう対応するかは任せますが、今のままでは不十分ですね。

maedana commented:
```
<%= link_to 'ログアウト', destroy_user_session_path, method: :delete %>
<%= link_to 'ユーザー一覧', users_path %>
<%= link_to 'プロフィール', users_my_profile_path %>
```
のように日本語を直接書いてしまっていてi18nで国際化されてないところが散見されます。i18nで国際化するようにしましょうー

Commit:
レビューに沿って修正させていただきました。

my_profileのルーティングを削除してuserのルーティングに統一
devise-i18nをインストールし国際化
不要なディレクトリやモジュール、コードの記載などを削除

maedana commented on app/assets/stylesheets/books.scss:
```diff
@@ -1,3 +0,0 @@
-// Place all the styles related to the books controller here.
-// They will automatically be included in application.css.
-// You can use Sass (SCSS) here: https://sass-lang.com/
```
不要なものを消しましょうって以前書いたんですが、一方で終了条件として

> サンプルアプリの既存の実装に「なんかおかしいな」と思ったところがあっても、そのまま放置で大丈夫です。プルリクエストには課題の要件満たすために必要なdiffだけ発生している状態がベストです。既存の実装を放置すべきか修正すべきか迷った場合はメンターに質問してください。

とあります。これはまさに上記にあたるもので削除することで余計なdiffが発生してしまってますね。PRの変更点は最低限になるように心がけるといいかなと思います。

kotokotodonabe replied:
なるほど。
不要な記述だと思ったので削除したのですが、余計なdiffを増やしてしまっているだけでした。

修正します🙇

maedana commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def create; end
+
+  def update; end
```
create/updateを本来のもの`Devise::RegistrationsController`のcreateやupdateをオーバーライドしていますが、中身が空っぽなので動作しなくなってますよ。ユーザ登録や更新がお手元で動いてないのではないでしょうか?

maedana commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,13 @@
+class Users::SessionsController < Devise::SessionsController
+  before_action :configure_sign_in_params, only: [:create]
+
+  def create; end
```
同上です。これではログイン出来なくなっているように見えます

maedana commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def configure_sign_in_params
+    devise_parameter_sanitizer.permit(:sign_in, keys: %i[email password])
+  end
```
こちらはカスタマイズの必要がないと思うんですよね。どういう意図で変えたんでしょうか?

kotokotodonabe replied:
@maedana 
ログイン時に不正な値を入力させないために、ストロングパラメータを追加させていただきました。

> カスタマイズの必要がない

ログイン時はストロングパラメータを記述する必要がないということでしょうか？

kotokotodonabe replied:
@maedana 
``` ruby
def configure_sign_in_params
    devise_parameter_sanitizer.permit(:sign_in, keys: %i[email password])
  end
```
上記のsign_inの箇所のソースコードを見てきました。

<img width="620" alt="スクリーンショット 2022-11-08 11 31 41" src="https://user-images.githubusercontent.com/80808445/200460749-482d1551-07e0-49cb-b93c-9e36ecd2239a.png">

上記のコードはデフォルトでpasswordとremember_meの値を許可しているという解釈でよろしいでしょうか？
そうなるとストロングパラメーターのkeysをカスタマイズして `:email` を追加する必要があるのかなと自分なりに考えてみたのですが、この辺りご教授いただきたいです。

何卒よろしくお願いいたします🙇

kotokotodonabe replied:
@maedana 
すいません、理解しました。
views/sessions/new.html.erbファイルをカスタマイズしていないので、不要でした。

ありがとうございました🙇

maedana commented on app/helpers/books_helper.rb:
```diff
@@ -1,4 +0,0 @@
-# frozen_string_literal: true
-
-module BooksHelper
-end
```
同上です。余計な差分が出ないのが望ましいです。

maedana commented on config/locales/devise.en.yml:
```diff
@@ -0,0 +1,78 @@
+# Additional translations at https://github.com/heartcombo/devise/wiki/I18n
+
+en:
+  users:
```
deviseのviewsのpathが`app/views/users`配下になっているので必要になってますが、devise標準の`app/views/devise`配下にしたらこのファイルなくせるかなと思います。

devise-i18nに入っている https://github.com/tigrish/devise-i18n/blob/master/rails/locales/en.yml がそのまま参照出来るようになるかなと。

maedana commented on config/locales/devise.views.ja.yml:
```diff
@@ -0,0 +1,153 @@
+        password: パスワード
+    models:
+      user: ユーザー
+  users:
```
同上です。deviseのviewsのpathが`app/views/users`配下になっているので必要になってますが、devise標準の`app/views/devise`配下にしたらこのファイルなくせるかなと思います。

devise関連以外で必要な翻訳(users#index, users#showに関するもの)は別ファイルにしてdevise関連とは分けるほうがいいかなと思います。

maedana commented on config/routes.rb:
```diff
@@ -1,4 +1,14 @@
+  devise_for :users, controllers: {
+    passwords: 'users/passwords',
+    registrations: 'users/registrations',
+    sessions: 'users/sessions'
```
controllerのところで言及しましたが、結局カスタマイズが必要なのは`registrations`だけじゃないかなって思うんですよね。いかがでしょうか?

maedana commented on test/controllers/my_profile_controller_test.rb:
```diff
@@ -0,0 +1,10 @@
+    get my_profile_show_url
+    assert_response :success
+  end
+end
```
こういうのも自動生成で使われてないファイルですねー。他にも不要なテスト関連のファイルがありますから消しましょう

maedana commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+<%= link_to t('users.defaults.sign_out'), destroy_user_session_path, method: :delete %>
+<%= link_to t('users.defaults.user_list'), users_path %>
+<%= link_to t('users.defaults.profile'), user_path(current_user.id) %>
```
細かいですが、`user_path(current_user)`でも大丈夫ですね

maedana commented on app/views/users/mailer/reset_password_instructions.html.erb:
```diff
@@ -0,0 +1,8 @@
+<p><%= link_to t('.action'), edit_password_url(@resource, reset_password_token: @token) %></p>
+
+<p><%= t('.instruction_2') %></p>
+<p><%= t('.instruction_3') %></p>
```
devise-i18nで生成して中身をカスタマイズしてないならば不要ですねー。他もカスマイズが入ってないならコミットする必要がないですね

Commit:
コントローラーのオーバーライドを再度定義した。

課題提出前にLintに該当箇所が抵触したので、修正した際に誤ってオーバーライドを削除した状態で提出してしまったため、それの修正を行いました。

Commit:
不要なdiffを発生してしまっていたため修正した。

Commit:
余計なdiffが出ていたファイルを修正しました。

Commit:
devise標準の翻訳を参照するため、viewのusersディレクトリをdeviseに名前を変更した。
 Please enter the commit message for your changes. Lines starting

Commit:
show.html.erbファイルのコードを変更や不要なファイルの削除

Commit:
sessionsコントローラーのルーティングが不要だったため、削除した。

Commit:
rubocopを通した。

maedana commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  # deviseのupdate時のストロングパラメータ
+  def configure_account_update_params
+    devise_parameter_sanitizer.permit(:account_update, keys: %i[email password post_code address self_introduction])
+  end
```
今回の対応でOKですが、別解として公式で紹介されているものは以下ですね。ご参考までに
https://github.com/heartcombo/devise#strong-parameters

kotokotodonabe replied:
@maedana 
ありがとうございます！！

maedana commented on config/locales/devise.views.ja.yml:
```diff
@@ -0,0 +1,147 @@
+      not_locked: はロックされていません。
+      not_saved:
+        one: エラーが発生したため %{resource} は保存されませんでした。
+        other: "%{count} 件のエラーが発生したため %{resource} は保存されませんでした。"
```
devise.en.yml にも言えることなんですが、これまでの対応で devise-i18nに入っている翻訳ファイルを基本そのまま使えばいい状態になったのではないでしょうか?

https://github.com/tigrish/devise-i18n/blob/master/rails/locales/ja.yml
https://github.com/tigrish/devise-i18n/blob/master/rails/locales/en.yml

つまり、devise用の独自翻訳ファイルは不要になったと思うんですよね。いかがでしょうか?

Commit:
devise用の独自翻訳ファイルが不要だったため、削除した。
